연락처 관리 앱 [Step1] 쫑, 록스

KimRoks opened the pull request:
리뷰어
@forestjae

---

팀원
@KimRoks, @shlim0

---

- 프로젝트 구조
<img width="298" alt="image" src="https://github.com/tasty-code/ios-contact-manager-ui/assets/113083860/63ebc263-14a4-4827-acb5-84cbda2e510f">

- protocols
    - `JSONCodable`:  `.json` 파일을 decoding 하는 코드를 따로 분리하고 재사용 하기 위해 생성한 protocol 입니다.
- Extenstions
    - `JSON+Codable`: `.json` 파일을 decoding 하는 코드를 따로 분리하기 위해 생성한 extension 입니다. `JSONCodable`을 채택하는 class에서 따로 구현할 필요 없이 하기 위해 해당 extension을 만들었습니다. Naming을 `JSON+Decodable`이 아닌`JSON+Codable`으로 한 이유는 추후 encoding을 고려하여 작성하였습니다.
- Models
    - `ContactJSON`: 연락처 정보가 있는 dummy `.json` 파일입니다.
    - `ContactDTO`: decoding 된`.json` 파일을 받을 모델입니다.
- Controllers
    - `ContactViewController`: 첫 번째 화면을 나타내는 VC 입니다. 연락처 정보가 없는 경우, cell을 불러오지 않습니다.

## 고민했던 점 / 궁금한 점

1. `textLabel?.text`이 deprecated 되었는데 아직 사용 가능한 이유는 무엇인가?
    
    → 저희 끼리 토의한 결과, 하위 호환성을 위해 아직 사용할 수 있도록 Apple이 남겨둔 것이라 생각했습니다. 혹시 다른 사실을 알고 계시다면 알려주세요!
    
    - `textLabel?.text` vs `defaultContentConfiguration`
        
- https://developer.apple.com/videos/play/wwdc2020/10027/
    - 해당 링크를 통해 WWDC 영상을 본 결과, `defaultContentConfiguration`을 통해 `TableViewCell` 뿐만 아니라 CollectionViewCell에서도 사용할 수 있는 것을 통해 호환성이 더 좋다고 느꼈습니다. 게다가, cell의 속성을 각각 설정하기 더 쉽다는 것도 알았습니다.
    - 또한, `defaultContentConfiguration`을 사용하면 cell의 새로운 configuration을 매번 만들어야 함에도 불구하고 `textLabel?.text`과 같이 configuration을 매번 만들지 않는 경우와 memory 사용량이 비슷하기 때문에 성능 향상이 되었다고 판단했습니다.
    - 그러나, 아직까지는 왜 `defaultContentConfiguration` 방식을 사용하는 것이 더 나은 선택인지 직관적으로 와닿지 않기 때문에 좋은 예시를 설명해 주시면 좋겠습니다.

2. `UITableViewCell()` vs `dequeueReusableCell()`
- cell을 매번 새로 생성하는 것 보다, view에 보이지 않는 cell을 queue에 두고 재사용 하는 것이 CPU 사용량 측면에서 더 나은 결과를 도출했습니다.

### **참고. cell 생성시 성능차이 비교**

> 1번과 2번 내용을 토대로 성능 비교를 해봤습니다.
>
- 실험 목적
    - `textLabel?.text`와 같이 cell에 정보를 표시하는 **Instance Property**가 iOS 14.0부터`deprecated` 되었고, `defaultContentConfiguration`으로 사용하라는 안내를 받았습니다.
- 실험 코드
    - 실험의 편의상 `!`(force unwrapping)을 사용했습니다.

```swift
func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
/// 1. UITableViewCell
        //  let cell = UITableViewCell()

/// 2. dequeueReusableCell (recommended)
        //  let cell: UITableViewCell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
      
      cell.accessoryType = .disclosureIndicator
      
/// a. textLabel?.text (deprecated until iOS 14.0)
      //  cell.textLabel?.text = ContactDTO![indexPath.row].name + "(" + ContactDTO![indexPath.row].age + ")"
      //  cell.detailTextLabel?.text = ContactDTO![indexPath.row].phoneNumber
      
/// b. defaultContentConfiguration (after iOS 14.0)
      //  var content = cell.defaultContentConfiguration()
      //  content.text = self.ContactDTO![indexPath.row].name + " (\(self.ContactDTO![indexPath.row].age))"
      //  content.secondaryText = self.ContactDTO![indexPath.row].phoneNumber
      //  cell.contentConfiguration = content
      return cell
}
```

- 실험 환경
    - Simulator가 아닌 실제 IPhone 12 pro 에서 실험했습니다. 아래 영상과 같은 방식으로 실험했습니다(실제 실험 영상 아님).

https://github.com/tasty-code/ios-contact-manager-ui/assets/113083860/92ecdbb5-b603-4978-80c8-57032c38abdc



- 실험 결과(왼쪽 상단) 1a, (오른쪽 상단) 1b, (왼쪽 하단) 2a, (오른쪽 하단) 2b

https://github.com/tasty-code/ios-contact-manager-ui/assets/113083860/d328aee7-cc2e-4a00-a80b-df64384bbd16




3. `.json` 파일을 decoding 하는 코드를 `ContactViewController`에서 분리하기 위해 별도로 protocol(`JSONCodable`)을 두고, 그 protocol을 extension(`JSON+Codable`) 하는 방식으로 파일을 구성했습니다. 이러한 방식으로 구성하는 것에 대해 리뷰어님의 개인적인 의견을 듣고 싶습니다.

4. Model(`ContactDTO`)이 아닌 `ContactViewController`에 `contactDTO`을 다루게 했습니다.
    
    ```swift
    private var contactDTO: [ContactDTO]?
    ```
    
    논리적으로, `contactDTO`을 Model(`ContactDTO`)에 두는 것이 옳다고 생각합니다. 저희는 `contactDTO`을 VC에서 작성하는것과 Model에 작성하는것 중 어느것이 더 적합한지 궁금합니다.

shlim0 committed:
feat: 연락처 모델 추가

shlim0 committed:
feat: JSON 더미 데이터 추가

KimRoks committed:
feat: JSON 더미 데이터 추가

shlim0 committed:
Merge remote-tracking branch 'origin/STEP1' into STEP1

shlim0 committed:
Revert "feat: JSON 더미 데이터 추가"

This reverts commit 145443ed045844639f3a9c15e4c9ba57b6ec309d.

shlim0 committed:
fix: git conflict 해결

shlim0 committed:
feat: cell의 accessoryType 추가

KimRoks committed:
refactor: 코드 개선

- JSON 코드 분리
- force unwrapping 제거
- 기타 네이밍 변경

KimRoks committed:
chore: class 이름에 맞게 파일 이름 변경

shlim0 committed:
chore: .DS_Store ignore 처리

shlim0 committed:
chore: 프로젝트 파일을 열 때 Model 파일이 존재함에도 존재하지 않는 문제 해결

forestjae commented on ContactManager/ContactManager/Controllers/ContactViewController.swift:
```diff
@@ -0,0 +1,41 @@
+final class ContactViewController: UIViewController, UITableViewDelegate, UITableViewDataSource, JSONCodable {
+    
+    @IBOutlet private weak var tableView: UITableView!
+    private var contactDTO: [ContactDTO]?
```
네이밍에 배열인게 드러나지 않아 조금 헷갈리네요

KimRoks replied:
반영하도록 하겠습니다!

forestjae commented on ContactManager/ContactManager/protocols/JSONCodable.swift:
```diff
@@ -0,0 +1,5 @@
+import Foundation
+
+protocol JSONCodable {
+    func decodeJSON(_ contactModel: inout [ContactDTO]?)
+}
```
👍 👍

forestjae commented on ContactManager/ContactManager/Extensions/JSON+Codable.swift:
```diff
@@ -0,0 +1,15 @@
+import UIKit
+
+extension JSONCodable {
+    func decodeJSON(_ contactModel: inout [ContactDTO]?) {
```
inout 키워드를 사용하신 이유가 무엇인가요?

KimRoks replied:
struct로 `ContactDTO` 작성되어 있고, 해당 인스턴스의 값을 바꿔주기 위해 사용했습니다.
기본적으로, 메서드의 파라미터는 `let` 이며 `ContactDTO`이 class가 아니기 때문에 `ContactDTO`를 return 하는 방식으로 작성할 수 없다고 생각했습니다.

답변을 생각하다 `inout` 키워드와 class를 사용하여 `ContactDTO`를 return 하는 것에 대해 성능차이가 있는지 궁금합니다.
저희가 알아보기론 메모리에 직접 접근하여 값을 바꾸기 때문에 성능적인 이점이 있을 것 으로 판단했습니다.

또한, `inout` 키워드를 사용하는 특정 상황 / 사용하지 않는 상황이 혹시 있을까요?
`inout` 키워드를 남용하면 코드의 복잡도(가독성)가 증가하고 `sideEffect`를 발생할 우려가 있다고 생각합니다.

forestjae replied:
말씀하신대로, 단점이 명확해서 꼭 필요한 상황이 아니면 사용하지 않는게 좋습니다.

구조체에 inout 키워드를 사용하게되면, 구조체가 가지는 성능상의 이점을 포기하게 되어 클래스와 큰 차이가 없게 됩니다.

shlim0 replied:
inout 키워드에 대한 리뷰에 대한 답변

WWDC 16 - Understanding Swift Performance를 참고하여 struct vs class의 성능을 비교해봤습니다. Struct의 경우 stack에 해당 struct의 instance를 생성하고 값의 변경이 일어날 때 heap 영역에 접근하는 것으로 알고 있습니다. 이 점을 잊고 struct로 진행 했던 것 같습니다. 따라서, class로 변경하는 것도 좋을 것 같습니다.

다만, DTO를 일반적으로 struct로 사용하는지, class로 사용하는지 숲재의 개인적인 의견이 궁금합니다. 일반적으로는 DTO를 struct로 둔다고 알고 있습니다. DTO의 취지상 값이 immutable 해야 한다고 생각합니다. (물론, inout 키워드를 이용하여 DTO를 변경하긴 합니다)

forestjae commented on ContactManager/ContactManager/Controllers/ContactViewController.swift:
```diff
@@ -0,0 +1,41 @@
+        cell.accessoryType = .disclosureIndicator
+        var content = cell.defaultContentConfiguration()
+        
+        guard let name = contactDTO?[indexPath.row].name,
+              let age = contactDTO?[indexPath.row].age,
+              let phoneNumber = contactDTO?[indexPath.row].phoneNumber 
```
contactDTO로 guard let을 한번만 해도 될 것 같습니다.

KimRoks replied:
지금처럼 ContactDTO에 프로퍼티가 몇 개 없는 경우, 해당 방식이 코드를 이해하는 데 더 직관적이라 생각합니다!
또한, 숲재가 얘기주신 것 처럼 ContactDTO에 프로퍼티가 많아졌을 때를 고려하면 말씀해주신 방식이 더 좋을 것 같습니다. 코드가 더 간결해지기 때문입니다. 

피드백 감사합니다. 수정했습니다!

```swift
//        guard let name = contactDTO?[indexPath.row].name,
//              let age = contactDTO?[indexPath.row].age,
//              let phoneNumber = contactDTO?[indexPath.row].phoneNumber 
//        else {
//            return UITableViewCell()
//        }

        guard let contactDTOs = contactDTO?[indexPath.row] else {
            return UITableViewCell()
        }
       
       // 더 간결해 졌습니다. 
        let name = contactDTOs.name
        let age = contactDTOs.age
        let phoneNumber = contactDTOs.phoneNumber
```

forestjae commented on ContactManager/ContactManager/Extensions/JSON+Codable.swift:
```diff
@@ -0,0 +1,15 @@
+            return
+        }
+        
+        guard let data = try? Data(contentsOf: path) else {
```
try? 와 try! 그리고 그냥 try의 차이점을 잘 알고 계시나요?

shlim0 replied:
그냥 `try`를 쓰는 경우는 `do-catch`를 작성하여 예외를 처리해야 합니다.
`try?`와 `try!`는 `do-catch`를 작성할 필요 없어 더 편리한 것으로 알고 있습니다.
`try?`의 경우, optional type이기 때문에 `if-let`이나 `guard-let`으로 optional binding 하는 경우, `nil`인 경우를 에러로 간주하고 넘어갑니다.
`try!`의 경우, `try?`와 같은 상황에 사용하나 force-unwrapping 하기 때문에 `nil`인 경우 run-time error가 발생하니 사용을 지양해야 한다고 알고 있습니다.

혹시 제가 모르는 내용이 있다면 코멘트 부탁드립니다.

forestjae commented:
고생 많으셨습니다! 고민을 많이 한 흔적이 보이네요 👏 👏 👏 

>   
>   * `JSONCodable`:  `.json` 파일을 decoding 하는 코드를 따로 분리하고 재사용 하기 위해 생성한 protocol 입니다.
> * Extenstions
>

재사용하기에는 특정 타입에 한정되어있어 보여요. 나중에 generic을 배우신다면 활용해서 재사용 가능한 코드를 만들 수 있으실 거에요.
   
>   * `JSON+Codable`: `.json` 파일을 decoding 하는 코드를 따로 분리하기 위해 생성한 extension 입니다. `JSONCodable`을 채택하는 class에서 따로 구현할 필요 없이 하기 위해 해당 extension을 만들었습니다. Naming을 `JSON+Decodable`이 아닌`JSON+Codable`으로 한 이유는 추후 encoding을 고려하여 작성하였습니다.
> * Models
>   
>   * `ContactJSON`: 연락처 정보가 있는 dummy `.json` 파일입니다.
>   * `ContactDTO`: decoding 된`.json` 파일을 받을 모델입니다.
> * Controllers
>   
>   * `ContactViewController`: 첫 번째 화면을 나타내는 VC 입니다. 연락처 정보가 없는 경우, cell을 불러오지 않습니다.
> 
> ## 고민했던 점 / 궁금한 점
> 1. `textLabel?.text`이 deprecated 되었는데 아직 사용 가능한 이유는 무엇인가?
>    → 저희 끼리 토의한 결과, 하위 호환성을 위해 아직 사용할 수 있도록 Apple이 남겨둔 것이라 생각했습니다. 혹시 다른 사실을 알고 계시다면 알려주세요!

네, 사용해 보시면 deprecated되었다고 warning이 뜰거에요. 나중에 애플이 내부적으로 코드를 바꾸면 크래시가 날 수 있기 때문에 사용하지 않아야 합니다.


>    
>    * `textLabel?.text` vs `defaultContentConfiguration`
> 
> * https://developer.apple.com/videos/play/wwdc2020/10027/
>   
>   * 해당 링크를 통해 WWDC 영상을 본 결과, `defaultContentConfiguration`을 통해 `TableViewCell` 뿐만 아니라 CollectionViewCell에서도 사용할 수 있는 것을 통해 호환성이 더 좋다고 느꼈습니다. 게다가, cell의 속성을 각각 설정하기 더 쉽다는 것도 알았습니다.
>   * 또한, `defaultContentConfiguration`을 사용하면 cell의 새로운 configuration을 매번 만들어야 함에도 불구하고 `textLabel?.text`과 같이 configuration을 매번 만들지 않는 경우와 memory 사용량이 비슷하기 때문에 성능 향상이 되었다고 판단했습니다.
>   * 그러나, 아직까지는 왜 `defaultContentConfiguration` 방식을 사용하는 것이 더 나은 선택인지 직관적으로 와닿지 않기 때문에 좋은 예시를 설명해 주시면 좋겠습니다.

지금은 defaultConfiguration만 사용해보셨지만, 나중에 좀 더 복잡한 셀을 만들기 위해 custom Configuration을 만드셔야 될 때가 있습니다. 그때 빛을 발하게 되는 방식입니다. 나중에 class와 struct의 성능차이를 배우게 되시면, 그때 확실히 이해하실 수 있을거에요.

> 
> 2. `UITableViewCell()` vs `dequeueReusableCell()`
> 
> * cell을 매번 새로 생성하는 것 보다, view에 보이지 않는 cell을 queue에 두고 재사용 하는 것이 CPU 사용량 측면에서 더 나은 결과를 도출했습니다.
> 
> ### **참고. cell 생성시 성능차이 비교**
> > 1번과 2번 내용을 토대로 성능 비교를 해봤습니다.
> 
> * 실험 목적
>   
>   * `textLabel?.text`와 같이 cell에 정보를 표시하는 **Instance Property**가 iOS 14.0부터`deprecated` 되었고, `defaultContentConfiguration`으로 사용하라는 안내를 받았습니다.
> * 실험 코드
>   
>   * 실험의 편의상 `!`(force unwrapping)을 사용했습니다.
> 
> ```swift
> func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
> /// 1. UITableViewCell
>         //  let cell = UITableViewCell()
> 
> /// 2. dequeueReusableCell (recommended)
>         //  let cell: UITableViewCell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
>       
>       cell.accessoryType = .disclosureIndicator
>       
> /// a. textLabel?.text (deprecated until iOS 14.0)
>       //  cell.textLabel?.text = ContactDTO![indexPath.row].name + "(" + ContactDTO![indexPath.row].age + ")"
>       //  cell.detailTextLabel?.text = ContactDTO![indexPath.row].phoneNumber
>       
> /// b. defaultContentConfiguration (after iOS 14.0)
>       //  var content = cell.defaultContentConfiguration()
>       //  content.text = self.ContactDTO![indexPath.row].name + " (\(self.ContactDTO![indexPath.row].age))"
>       //  content.secondaryText = self.ContactDTO![indexPath.row].phoneNumber
>       //  cell.contentConfiguration = content
>       return cell
> }
> ```
> 
> * 실험 환경
>   
>   * Simulator가 아닌 실제 IPhone 12 pro 에서 실험했습니다. 아래 영상과 같은 방식으로 실험했습니다(실제 실험 영상 아님).
> 
>  default.mp4 
> * 실험 결과(왼쪽 상단) 1a, (오른쪽 상단) 1b, (왼쪽 하단) 2a, (오른쪽 하단) 2b
> 
>  default.mov 

이 실험의 목적은 무엇인가요?


> 3. `.json` 파일을 decoding 하는 코드를 `ContactViewController`에서 분리하기 위해 별도로 protocol(`JSONCodable`)을 두고, 그 protocol을 extension(`JSON+Codable`) 하는 방식으로 파일을 구성했습니다. 이러한 방식으로 구성하는 것에 대해 리뷰어님의 개인적인 의견을 듣고 싶습니다.

위에 JSON 관련해서 남긴 코멘트 확인해보세요!


> 4. Model(`ContactDTO`)이 아닌 `ContactViewController`에 `contactDTO`을 다루게 했습니다.
>    ```swift
>    private var contactDTO: [ContactDTO]?
>    ```
>      
>    논리적으로, `contactDTO`을 Model(`ContactDTO`)에 두는 것이 옳다고 생각합니다. 저희는 `contactDTO`을 VC에서 작성하는것과 Model에 작성하는것 중 어느것이 더 적합한지 궁금합니다.

DTO 정의는 지금처럼 Model 폴더에 두는게 맞아보여요. 그런데 contactDTO 변수를 VC가 알아야 테이블 뷰를 그리지 않나요? VC에 아예 두지 않으면 받아올 방법이 없지 않나요? 질문이 잘 이해가 안되네요 😂

forestjae commented on ContactManager/ContactManager/Extensions/JSON+Codable.swift:
```diff
@@ -0,0 +1,15 @@
+import UIKit
+
+extension JSONCodable {
```
JSON + Codable로 네이밍 하신 이유가 궁금해요.

일반적으로 "원래의 파일명" + "extension에서 추가한 기능이나 프로토콜의 이름" 으로 명명합니다.

shlim0 replied:
이건 저희가 네이밍 규칙을 잘못 알고 있었던 것 같습니다. 이미 프로토콜로 `JSONCodable.swift`이 존재하는데, extension으로 `JSONCodable.swift`가 있을 순 없기 때문에 `JSON+Codable.swift` 인 줄 알았습니다. `JSONCodable+decodeJSON.swift` 으로 수정했습니다.

shlim0 committed:
refactor: 파라미터 이름을 contactModel에서 contactDTOcontactDTO 변경

shlim0 committed:
refactor: 네이밍 및 binding 구조 개선

1. contactDTO 네이밍을 contactDTOs로 변경
2. contactDTO 프로퍼티를 하나씩 optional binding 하는 것에서, contactDTO만 optional binding 하도록 함

shlim0 committed:
chore: 한 줄 제거

shlim0 commented:
> UITableViewCell() vs dequeueReusableCell()
> 
> - cell을 매번 새로 생성하는 것 보다, view에 보이지 않는 cell을 queue에 두고 재사용 하는 것이 CPU 사용량 측면에서 더 나은 결과를 도출했습니다.
> 
> ### **참고. cell 생성시 성능차이 비교**
> 
> > 1번과 2번 내용을 토대로 성능 비교를 해봤습니다.
> > 
> - 실험 목적
>     - `textLabel?.text`와 같이 cell에 정보를 표시하는 **Instance Property**가 iOS 14.0부터`deprecated` 되었고, `defaultContentConfiguration`으로 사용하라는 안내를 받았습니다.
> - 실험 코드
>     - 실험의 편의상 `!`(force unwrapping)을 사용했습니다.
> 
> ```swift
> func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
> /// 1. UITableViewCell
>         //  let cell = UITableViewCell()
> 
> /// 2. dequeueReusableCell (recommended)
>         //  let cell: UITableViewCell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
> 
>       cell.accessoryType = .disclosureIndicator
> 
> /// a. textLabel?.text (deprecated until iOS 14.0)
>       //  cell.textLabel?.text = ContactDTO![indexPath.row].name + "(" + ContactDTO![indexPath.row].age + ")"
>       //  cell.detailTextLabel?.text = ContactDTO![indexPath.row].phoneNumber
> 
> /// b. defaultContentConfiguration (after iOS 14.0)
>       //  var content = cell.defaultContentConfiguration()
>       //  content.text = self.ContactDTO![indexPath.row].name + " (\\(self.ContactDTO![indexPath.row].age))"
>       //  content.secondaryText = self.ContactDTO![indexPath.row].phoneNumber
>       //  cell.contentConfiguration = content
>       return cell
> }
> 
> ```
> 
> - 실험 환경
>     - Simulator가 아닌 실제 IPhone 12 pro 에서 실험했습니다. 아래 영상과 같은 방식으로 실험했습니다(실제 실험 영상 아님).
> 
> default.mp4
> 
> - 실험 결과(왼쪽 상단) 1a, (오른쪽 상단) 1b, (왼쪽 하단) 2a, (오른쪽 하단) 2b
> 
> default.mov
> 
> > 이 실험의 목적은 무엇인가요?
> > 

→ UITableViewCell() vs dequeueReusableCell() 와 함께 `textLabel?.text` vs `defaultContentConfiguration` 과 같은 방식으로 사용했을 때 (총 4가지 경우) 어떤 것이 더 성능에 이점이 있는지 확인하기 위해 실험했습니다. 또, 직관적으로 어떤 것이 더 나은 선택지인지 파악하기 위해 이른바 ‘삽질’을 해보려 했던 것 같습니다.

> Model(ContactDTO)이 아닌 ContactViewController에 contactDTO을 다루게 했습니다.private var contactDTO: [ContactDTO]?
논리적으로, contactDTO을 Model(ContactDTO)에 두는 것이 옳다고 생각합니다. 저희는 contactDTO을 VC에서 작성하는것과 Model에 작성하는것 중 어느것이 더 적합한지 궁금합니다.
> 
> 
> > DTO 정의는 지금처럼 Model 폴더에 두는게 맞아보여요. 그런데 contactDTO 변수를 VC가 알아야 테이블 뷰를 그리지 않나요? VC에 아예 두지 않으면 받아올 방법이 없지 않나요? 질문이 잘 이해가 안되네요 😂
> > 

→ Model 폴더에 contactDTO 변수를 두고 Model 폴더에서 contactDTO 변수를 넘겨주는 getter를 만드는 방식으로 VC에 넘겨주는 걸 생각하고 작성했습니다!

shlim0 committed:
chore: 파일 이름 변경 반영

KimRoks committed:
refactor: ContactVC의 프로토콜 분리
